src/PAO_grid_transform_module.f90

```diff
@@ -115,127 +115,147 @@ subroutine single_PAO_to_grid(pao_fns)
-                rcut = r_h + RD_ERR
-                call check_block (xblock,yblock,zblock,xatom,yatom,zatom, rcut, &  ! in
-                     npoint,ip_store,r_store,x_store,y_store,z_store,n_pts_in_block) !out
-
-                if(npoint > 0) then
-                   current_num_blocks = npao_species(the_species)*n_pts_in_block
-                   temp_block_storage = zero
-                   offset_position = no_of_ib_ia
-                   !$omp parallel do default(none) &
-                   !$omp             schedule(dynamic) &
-                   !$omp             reduction(+: temp_block_storage) &
-                   !$omp             shared(n_pts_in_block, pao, the_species, offset_position, &
-                   !$omp                    npoint, no_of_ib_ia, ip_store, r_store, x_store, y_store, z_store) &
-                   !$omp             private(r_from_i, ip, position, ipoint, &
-                   !$omp                     x, y, z, l1, acz, m1, count1, val)
-                   do ip=1,npoint
-                      ipoint=ip_store(ip)
-                      position= offset_position + ipoint
-
-                      r_from_i = r_store(ip)
+                call check_block (xblock(iblock), yblock(iblock), zblock(iblock), &
+                     xatom(ia, ipart, iblock), &
+                     yatom(ia, ipart, iblock), &
+                     zatom(ia, ipart, iblock), &
+                     rcut, &  ! in
+                     npoint, ip_store, r_store, x_store, y_store, z_store, & !out
+                     n_pts_in_block) ! in
+
+                my_species = species_store(ia, ipart, iblock)
+
+                npoint_if_omp : if(npoint > 0) then
+                   points_loop_omp: do ip=1,npoint
+                      position = offset_position(ia, ipart, iblock) + ip_store(ip)
                       x = x_store(ip)
                       y = y_store(ip)
                       z = z_store(ip)
                       ! For this point-atom offset, we accumulate the PAO on the grid
-                      count1 = 1
-                      do l1 = 0,pao(the_species)%greatest_angmom
-                         do acz = 1,pao(the_species)%angmom(l1)%n_zeta_in_angmom
-                            do m1=-l1,l1
-                               call evaluate_pao(the_species,l1,acz,m1,x,y,z,val)
-                               !gridfunctions(pao_fns)%griddata(position+(count1-1)*n_pts_in_block) = val
-                               temp_block_storage(ipoint + (count1-1)*n_pts_in_block) = val
+                      count1 = 0
+                      l_loop: do l1 = 0,pao(my_species)%greatest_angmom
+                         z_loop: do acz = 1,pao(my_species)%angmom(l1)%n_zeta_in_angmom
+                            m_loop: do m1=-l1,l1
+                               call evaluate_pao(my_species,l1,acz,m1,x,y,z,val)
+                               gridfunctions(pao_fns)%griddata(position + count1 * n_pts_in_block) = val
```

Comment:
We could remove count1 here and just increment position by n_pts_in_block at each loop step I think

Reply:
Well spotted! That makes it a lot clearer